### ISSUE TITLE
Consider upgrading to jsonschema 3.0

### ISSUE THREAD

**Issue by drdavella** [spacetelescope/asdf#530]
Currently there's only an alpha release, and it causes some errors that are detected by our unit tests. However, once things become more stable we should consider whether we should upgrade, and if so, what changes will need to be made.

Version 3.0 introduces support for [Draft-06](http://json-schema.org/draft-06/json-schema-release-notes.html). This could potentially allow us to update the ASDF Standard to support Draft-06 as well. However, if we are really serious about providing/supporting multiple implementations of ASDF, we need to look at the lowest common denominator across JSON Schema implementations, which might still be Draft-04.

**Comment by drdavella** [comment]
No, there's no urgency. I just wanted to record the issue here as a reminder for future reference.

**Comment by perrygreenfield** [comment]
Agreed. We should evaluate it once is it a stable release.

**Comment by drdavella** [comment]
Yes, it does seem very useful. I believe it is supported as of Draft-07. The `jsonschema` Python module doesn't yet support that draft (and only supports Draft-06 in alpha releases), but it looks like Draft-07 may be supported relatively shortly after that.

Even after that point, moving to Draft-07 may disadvantage other potential implementations of ASDF since JSON Schema support in other languages may not be as advanced. We'll need to have a discussion about how much that really matters, though.

**Comment by pllim** [comment]
Somehow 3.0 get picked up by one of my conda install that happened before I installed `asdf`. This caused conflict and to solve that:

* `conda uninstall jsonschema`

Then, I tried `pip install jsonschema` and that picked up 2.x but that did **not** work either. `asdf` gave a cryptic message about cannot understand something when I load a file. What finally worked:

* `pip uninstall jsonschema`
* Install `asdf` from `master` and let `asdf` install `jsonschema` in the process

I would say this kind of installation hurdle is not very user-friendly right now. I spent 30 minutes trying to make this work. FYI.

**Comment by pllim** [comment]
"Cannot understand something something 1.3.0"

I tried to reproduce it in a clean env but I cannot now. I swear it was happening last night. @ejeschke can be my witness!

The fact that `asdf` installs `jsonschema` if not there is nice, *but* it wasn't playing nice with `jsonschema` that is already installed (for one reason or another).

**Comment by drdavella** [comment]
I can't reproduce it either. I realize it's annoying to have a strict dependency on `jsonschema`, but that's just the situation we have for the near term. Maybe I can try to address this for 2.3.0, but it also depends on which versions of Python `jsonschema-3.0` supports. Also the fact that the latest released version of `jsonschema` on pypi is 2.6.0 means that I don't feel a strong need to do this immediately.

**Comment by drdavella** [comment]
Given the error message snippet you reported, it sounds like you were probably trying to read a file that was recently produced by the JWST pipeline. Unfortunately, since they run against a dev version of ASDF, it means that you won't be able to use a stable version to read their files. So installing from `master` was the right thing to do.

**Comment by Cadair** [comment]
It's worth noting that the `jsonschema 3.0.1` package is now up on PyPI and the like.

**Comment by drdavella** [comment]
There's actually a 3.0 release of `jsonschema` now so this has become more urgent/relevant.

**Comment by pllim** [comment]
`jsonschema<3` now gives this warning:

```
DeprecationWarning: Using or importing the ABCs from 'collections' instead of
from 'collections.abc' is deprecated, and in 3.8 it will stop working
```

It has been fixed in `jsonschema>=3`. Indirectly, this means that if `asdf` does not support `jsonschema>=3`, `asdf` will not work in Python 3.8.

**Comment by pllim** [comment]
Might still be a problem, see https://github.com/spacetelescope/asdf/pull/684#issuecomment-515732325

**Comment by jdavies-st** [comment]
This is an order-of-installation issue, not a problem in `asdf`.

Fixed by #3705.
